### PR TITLE
Remove useless cast nodes in graph after converting to fp16 model

### PR DESCRIPTION
**Pull request by xiaowuhu**
1. For ort 1.17, ORT will not remove cast(to16) automatically when meeting cast(to16)->cast(to32) pair. So we need remove this kind of pair in converter.
2. Due to security issue, onnx 1.15 and before cannot be used. So upgrade to 1.16 for all testing.

### CODE REVIEW

File: onnxconverter_common/float16.py [hunk 1]
```diff
@@ -388,3 +389,104 @@ def sort_topology(graph_proto):
...
+    cast_node_downstream_dict = {}   # mapping cast node(name) to its downstream node
+    for current_node in graph_proto.node:
+        # find the downstream node(s)
+        for input_name in current_node.input:
+            if input_name in output_name_to_cast_node_dict:
+                # found the downstream node of the cast node, might be multiple
+                cast_node = output_name_to_cast_node_dict[input_name]
+                if cast_node.name not in cast_node_downstream_dict:
+                    cast_node_downstream_dict[cast_node.name] = current_node
+                    #name_to_node_dict[current_node.name] = current_node
```
Comment (fatcat-z) [inline onnxconverter_common/float16.py:423]: Let's remove these comments in the final PR.

File: onnxconverter_common/float16.py [hunk 2]
```diff
@@ -388,3 +389,104 @@ def sort_topology(graph_proto):
             for g in attr.graphs:
                 if isinstance(g, onnx_proto.GraphProto):
                     sort_topology(g)  # sort sub-graph
+
+
+def remove_unnecessary_cast_node(graph_proto):
```
Comment (fatcat-z) [inline onnxconverter_common/float16.py:394]: Is there a way to create a test to cover the issue this function resolves?

**Review by fatcat-z — APPROVED** [review]

File: onnxconverter_common/float16.py [hunk 3]
```diff
@@ -388,3 +389,104 @@ def sort_topology(graph_proto):
...
+        # find the downstream node's input as second_cast_node's output
+        for i, input_name in enumerate(downstream_node.input):
+            for output_name in second_cast_node.output:
+                if input_name == output_name:
+                    # change the input as the upstream node's output
+                    downstream_node.input[i] = out
+
+    # 6. remove the cast node pair
+    for cast_node_pair in remove_candidate:
+        first_cast_node = cast_node_pair[0]
```
[the remark this answers is not in the document]
Comment (xiaowuhu) [reply onnxconverter_common/float16.py:489]: fixed

File: onnxconverter_common/float16.py (hunk 2 above)
Comment (xiaowuhu) [reply onnxconverter_common/float16.py:394]: The test case failed so have to change this code.

**Pull request merged**